src/settings.js

```diff
@@ -901,6 +902,8 @@ var BUNDLED_CD_DEBUG_FILE = ""; // Path to the CyberDWARF debug file passed to t
 
 var TEXTDECODER = 1; // Is enabled, use the JavaScript TextDecoder API for string marshalling.
                      // Enabled by default, set this to 0 to disable.
+var EMBIND_STD_STRING_IS_UTF8 = 1; // Embind specific: If enabled, assume UTF-8 encoded data in std::string binding.
+                                   // Disable this to support binary data transfer.
```

Comment:
@chadaustin wrote at may 24:

> std::string is ultimately an uninterpreted array of char - some people use it to contain binary data, some people use it to contain specific things like UTF-8-encoded data. JavaScript String is an array of UTF-16 code units. So the notion of binary data being stored in JavaScript String is already suspect.

> That said, and I don't have an Emscripten installation to test with right now, but it looks like Embind currently maps String to std::string by enumerating UTF-16 code units and disallowing any > 255. There may be existing users that rely on this behavior.

> I'd support adding either a compile-time flag (or even initialization-time flag) that changes this behavior and maps String <-> std::string as UTF-8. I'm not very comfortable changing the existing behavior without a flag, however.

So we simply do not know what data we get and could break existing code by change the mapping.

Reply:
Thanks, I missed that. So that is a crucial issue here - does the std::string contain any internal nulls, and is it terminated by a null.

I believe in c++11 null-termination is guaranteed, and embind requires c++11 anyhow, so that part is fine? But that leaves internal nulls.

Are we talking about conversion from a JS String to a c++ std::string, or the other direction - or both?

Reply:
We are talking about both directions in the mapping here. But there is more to it than embedding nulls and null termination.

According to chad, earlier in this thread, existing projects might use the std::string <> JS string mapping for pure binary data transfer(including embedded nulls). In this specific use case, UTF-8 encoding/decoding in general would be counterproductive as it would lead to garbled data on one or the other end. Hence the need for the EMBIND_STD_STRING_IS_UTF8 switch. It can be set to 0 to enable this "binary safe" mapping. In the (presumably much more common) use case, people can expect that UTF-8 data in std::string on C++ side would just be converted to JS string back and forth.

If you look at the latest issue on commit 2d4c7e5 (sizeInBytes parameter), it became obvious that during string conversion in embind, that data is not in null terminated form on the heap, hence the need for the sizeInBytes parameter in preamble/UTF8ArrayToString(). This point affects the C++>JS direction only.

@kripken and @chadaustin from our point of view we think we have, with some twists and turns, completed what can be done here. If it is ok for you, we would now like to concentrate on how we can merge/close this PR. UTF-8 error handling can be implemented on a separate request as you proposed.

Reply:
I'm fine with the changes as written and have no reason to hold up landing, beyond anything @kripken may still have. I think eventually it might make sense to decouple embind from preamble's UTF-8 stuff and do our own conversion and validation, but that doesn't need to hold this up.

Reply:
Thanks chad for your positive feedback. Yes embind could have separate UTF-8 conversion with validation. One challenge I can see there is how to avoid duplication with preamble.js, especially if branching/complexity in preamble should not deteriorate. 

As we are not familiar on how to finalize this, can you give some hints if there is something to do from our side? Does @kripken initiate the merge and decide the release schedule?

Reply:
I apologize, but I do still have a concern here. It feels odd to me to have a flag for the behavior of a core type like `std::string`. Tutorial examples would need to mention the two modes, etc.

I realize that changing the behavior here would be a breaking change, but I think it would be worth it actually. What I mean is, we'd say that embind maps std::string to JS Strings, and it assumes the std::strings are "regular" null-terminated strings (as c++11 ones are in their primary storage, in fact). That is the most natural mapping of std::string to JS Strings, I think. And for users that actually want arbitrary binary data in std::string, instead there is `std::vector<char>` which makes more sense there anyhow?

The risk of the breaking change can be mitigated by documenting this in the Changelog and in the main embind docs.

Another benefit of doing it that way would be avoiding the extra sizeInBytes param.

Again, I apologize for raising this now, as this PR has seen quite a lot of work. But this does seem like a crucial issue that I didn't realize until my last few questions. I think, though, that if there is agreement on this, then it should be a small amount of changes (just simplification) and we can get that landed soon I hope.

Reply:
I still think we should leave the option in (even though it defaults to UTF-8 now). embind doesn't currently support mapping JS binary data to vector<char>, for example. And I believe that, without leaving the option in, IMVU's usage would break. Perhaps @jlee-imvu could confirm.

Either way, we still need to support embedded nulls in the string.

Reply:
I'd be curious to hear about existing usage that would break. Also, whether those usages could be replaced by `std::vector<char>` - you say embind doesn't support JS binary => vector, what happens when you give embind a JS Array (is that not supported)?

(And why is embedded nulls something we need anyhow, separately from the above?)

Reply:
My memory is pretty fuzzy at this point and I obviously don't have source code to check, but I believe, at IMVU, we sometimes passed around strings of binary data in JS for API reasons. Vague recollections of XMLHttpRequest...

On the C++ side, std::string was the convention for buffers of untyped data. (This also appears to be a weak convention in my little corner of Facebook too.)

It's probably possible to have bind.cpp automatically register some sort of special mapping from String to std::vector<char>, but I don't immediately know what that would look like, so I'd expect it to be a nontrivial amount of work. Embind does support mapping arrays to std::vector in general, though.

In an ideal world, I think it makes sense for String -> std::string to UTF-8-encode the JS data into the C++ string value and for std::string -> String to UTF-8-decode (throwing an exception or aborting the process or something if the std::string contains invalid UTF-8), which is why I support defaulting to the new behavior. Especially since typed arrays will still map to std::string without interpretation. But I don't want to break existing use cases without providing at least an option to go back to the old behavior, at least as part of this change.

Maybe we should ping the community in six months or a year and see if it makes sense to remove the setting.

Your point about the docs is a good one though. @innova-engineering can you do a quick pass through any relevant embind docs and add any clarifying comments that make sense?

Reply:
Before we start investing further work in documentation changes, please let us know know what your decision about the EMBIND_STD_STRING_IS_UTF8 switch is.

Do you want to remove it, resulting in "std::string is always interpreted as UTF-8" or do you want to keep it (as currently implemented)?

One important point to note @kripken about the sizeInBytes param: Even if we remove EMBIND_STD_STRING_IS_UTF8, the sizeInBytes param can not be removed. As mentioned earlier: "data is not in null terminated form on the heap, hence the need for the sizeInBytes parameter". More specifically, data format on the heap during conversion currently is like this: <length in bytes (int)><data without null terminator>.
We even did the removal of sizeInBytes in commit 2d4c7e5 already, resulting in several broken units tests. Hence the reverse commit d448d1e.

Reply:
Ok, fair point about backwards compatibility, and makes sense to add the option and perhaps consider removing it later.

Then my last concern is that sizeInBytes param. This would increase code size for our default JS output, not just for embind-using code, and we've been trying to go in the other direction. Two thoughts:

 * First, I am probably missing something but it seems odd to me that `UTF8ArrayToString` is involved here. On the one hand, if this is just arbitrary binary data, then we should read the proper amount of bytes based on the size, and create a JS Array from that. If, on the other hand, this is a null-terminated string, then we can create a JS String from that. In both cases it doesn't seem like we need sizeInBytes?
 * In the worst case, we may want to create a new additional function, which has sizeInBytes (and maybe error handling etc.), and that would only be used with embind.

Reply:
UTF8ToString is the real function being modified, but it just delegates to UTF8ArrayToString, if I'm reading it correctly.

For the record, I'm personally fine with a variant of UTF8ToString that takes a size parameter and is only called by embind (or simply copying the new implementation into embind itself).

Reply:
Ok, cool, sounds like we agree on creating a new function to avoid increasing code size in the general case. Yeah, maybe it could be just in the embind code itself.

But I'd still like to understand why this is needed, something doesn't make sense to me. Maybe this is more of a question for @innova-engineering: It seems like we should have one of the following:

 * std::string is UTF8 data, null-terminated strings. We can convert them to a JS String and back. We don't need to know the size in bytes here.
 * std::string is raw binary data, just some amount of bytes. This can have 0s in the middle, and may not have a 0 at the end - it's just raw binary. We shouldn't convert this to a JS String (as that can't represent arbitrary raw binary), instead it should be a JS Array or Typed Array. In such conversions, we should not even reach any UTF8* function, so again, no need for sizeInBytes.

Where am I understanding this wrong?

Reply:
0 is a legal Unicode code point, meaning you can have a std::string containing valid UTF-8 that also contains embedded nul characters.

Also, String can represent raw binary: you just use code units in [0, 255] for each byte. IIRC, that's how people handled binary data in IE prior to typed arrays. :)

Reply:
For example, this is a legal JSON document: ["\u0000\u0000\u0000"]. If it were parsed and then its contents round-tripped through embind somehow, the string should continue to have length 3.

Reply:
Ok good, in trying to understand both of your concerns I've put up another commit:
The approach here is to use the original version of UTF8ArrayToString without sizeInBytes from embind, while still supporting embedded nulls.
Also the non-null termination issue needed some special handling in emind: null termination is now temporarily ensured on the heap during conversion.
For embedded nulls, UTF8ArrayToString is effectively called multiple times.

With this approach we have:
 * support for embedded nulls (as required by existing unit tests)
 * usage of the runtime and space optimized original UTF8ArrayToString version from embind
 * in the common case when there are no embedded nulls, and null termination might be already present, the performance impact in embind should be negligible (@chadaustin)?
 * no direct duplication of a somewhat medium-sized non-obvious algorithm in the codebase, as this move would literally cause myself pain... :-)

What I would still question now if we really need support for embedded nulls *in UTF-8 mode*? We do not break it in binary mode, and if we do not comprehensibly need it in UTF-8, 
I'd be happy to add this simplification. Furthermore, null termination handling and embeddedd null handling have been commented as non-obvious point in embind.

Reply:
My guess is that looping over calls to UTF8ToString could materially affect performance, especially given that std::string is a pretty common thing to marshal in hot loops. In particular, it makes multiple passes over the input data to detect embedded nulls. It makes me uncomfortable - I'd rather just have the sizeInBytes parameter, personally, but a duplicate function is okay too. But if looping over UTF8ToString calls is what it takes to make @kripken comfortable and get this thing landed, I'm okay with it.

Reply:
In an attempt bring some objectivity to possible performance issues, I've timed a modified embind unit test for string conversion. More specifically, I've prepared a ~3Mb sized string (above that, I was hitting some memory limit) and corresponding byte array. In the string, every third byte is a null-byte. I've used the emval_test_take_and_return_std_string() function from embind.test.js to trigger the HEAP->UTF-8 decode->JSstring conversion. In addition, I've hacked the EMBIND_STD_STRING_IS_UTF8 switch to select between the current implementation and a separate UTF8ArrayToString() function with the additional sizeInBytes param which I have re-added from a previous commit. Also, the TEXTDECODER path has been deactivated.

After several runs using "python tests/runner.py other.test_embind" on my system, I observe that this conversion takes between 300 and 350 ms, depending e.g. on optimization setting (-O1/-O2) and minor differences between individual runs. Between the implementations, I cannot measure any qualifyable performance advantage of the separate conversion function with sizeInBytes over the the current implementation which uses one pass to find embedded nulls and converts the individual parts with the existing UTF8ArrayToString.

Hoping this makes you a bit more more comfortable about it.

Reply:
300 ms is extremely large - how many times did the benchmark loop pass the string between JS and C++? In a realtime application, that would drop something like 20 frames. I wonder if the JIT wasn't warm yet?

Reply:
Conversion time grew (roughly linear) with data size. As said, this short test was running with ~3Mb (as large as I could get). The idea with the large test data was to minimize side effect differences. What I added now is a local loop over emval_test_take_and_return_std_string() and average conversion times calculation. Per unit test configuration I looped 20x locally, which resulted in 160 calls in total. Results are
 * [Current implementation]:
    avg across 160 calls: 332.07 ms
 * [Separate conversion function with sizeInBytes]:
    avg across 160 calls: 373.01 ms
    
I did this multiple times with very similar results. The tests were all run within node via the unit test suite. It even shows a slight advantage of the current implementation over the separate conversion function. Maybe the performance advantage we get from using the unmodified UTF8ArrayToString() outweights the additional "check for nulls" pass.

Reply:
OK. Those performance numbers still seem funny to me - passing a 3 MB string through embind should not take the same amount of time as a video game could render New York 20 times :) - but I'm okay with landing this diff as is. It's got the behavior we want and it's got tests, and all of the implementation details can be improved later if needed.

@kripken ?

Reply:
I agree, this looks good to land since it has the right behavior + tests. If we see perf issues later we can consider optimizing then at that time.